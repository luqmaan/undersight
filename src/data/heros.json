[
  {
    "name": "Soldier: 76",
    "role": "Offense",
    "icon": "soldier.png"
  },
  {
    "name": "Reaper",
    "role": "Offense",
    "icon": "reaper.png"
  },
  {
    "name": "Pharah",
    "role": "Offense",
    "icon": "pharah.png"
  },
  {
    "name": "Genji",
    "role": "Offense",
    "icon": "genji.png"
  },
  {
    "name": "McCree",
    "role": "Offense",
    "icon": "mccree.png"
  },
  {
    "name": "Tracer",
    "role": "Offense",
    "icon": "tracertracer.png"
  },
  {
    "name": "Bastion",
    "role": "Defense",
    "icon": "bastion.png"
  },
  {
    "name": "Hanzo",
    "role": "Defense",
    "icon": "hanzo.png"
  },
  {
    "name": "Junkrat",
    "role": "Defense",
    "icon": "junkrat.png"
  },
  {
    "name": "Mei",
    "role": "Defense",
    "icon": "mei.png"
  },
  {
    "name": "Torbjörn",
    "role": "Defense",
    "icon": "torbjorn.png"
  },
  {
    "name": "Widowmaker",
    "role": "Defense",
    "icon": "widowmaker.png"
  },
  {
    "name": "D.Va",
    "role": "Tank",
    "icon": "dva.png"
  },
  {
    "name": "Reinhardt",
    "role": "Tank",
    "icon": "reinhardt.png"
  },
  {
    "name": "Roadhog",
    "role": "Tank",
    "icon": "roadhog.png"
  },
  {
    "name": "Winston",
    "role": "Tank",
    "icon": "winston.png"
  },
  {
    "name": "Zarya",
    "role": "Tank",
    "icon": "zarya.png"
  },
  {
    "name": "Ana",
    "role": "Support",
    "icon": "ana.png"
  },
  {
    "name": "Lúcio",
    "role": "Support",
    "icon": "lucio.png"
  },
  {
    "name": "Mercy",
    "role": "Support",
    "icon": "mercy.png"
  },
  {
    "name": "Symmetra",
    "role": "Support",
    "icon": "symmetra.png"
  },
  {
    "name": "Zenyatta",
    "role": "Support",
    "icon": "zenyatta.png"
  }
]
