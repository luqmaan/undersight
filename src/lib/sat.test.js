import Logic from 'logic-solver';
import flatten from 'lodash/flatten';
import uniq from 'lodash/uniq';
import difference from 'lodash/difference';
import sortBy from 'lodash/sortBy';

import counters from '../data/counters.json';
import heros from '../data/heros.json';

function getTeamPicks(teamPicks, competitive) {
  const allHeros = heros.map((hero) => hero.name);

  const teamHardCounters = uniq(flatten(teamPicks.map((heroName) => {
    return counters
      .filter((counter) => counter.you === heroName)
      .filter((counter) => counter.score === 2)
      .map((counter) => counter.enemy);
  })));

  const uncountersHeros = difference(allHeros, teamHardCounters);
  const notPicked = difference(allHeros, competitive ? teamPicks : []);

  const scores = notPicked.map((heroName) => {
    const numHardCounters = counters.reduce((prev, counter) => {
      if (counter.you === heroName && counter.score === 2 && uncountersHeros.indexOf(counter.enemy) !== -1) {
        return prev + 1;
      }
      return prev;
    }, 0);

    return {
      score: numHardCounters,
      heroName,
    };
  });

  return sortBy(scores, 'score').reverse();
}


describe('sat', () => {
  fit('should asdf', () => {
    // T = your team
    // A = heroes currently hard countered by your team picks
    // find the hero not currently picked, who has the most hard counters who are not in a
    //
    // uncountersHeros = all heros - A
    // notPicked = all heros - T
    // hero with the most hard counters in uncountersHeros
    //
    // heros = []
    //
    // for hero in T:
    //   hard_ctrs = 0
    //   for ctr in counters:
    //     if ctr.you == hero and ctr.score == -2 and ctr.enemy in uncountersHeros:
    //       hard_ctrs += 1
    //   heros.append((hero, hard_ctrs))
    //
    // heros = sorted(cmp=lambda a, b: a[1] - b[1])
    //
    // num_hard_ctrs = counters[(counters.you == hero) & (counters.score == -2) & (counters.enemy in uncountersHeros)]
    // if num_hard_ctrs > max_hard_ctrs:
    //   current_pick = hero

    expect(getTeamPicks(['Pharah', 'D.Va', 'D.Va', 'D.Va', 'D.Va', 'D.Va'], false)).toMatchSnapshot();

  });

});
