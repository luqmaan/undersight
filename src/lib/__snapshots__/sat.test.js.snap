exports[`sat should asdf 1`] = `
Array [
  Object {
    "heroName": "Mei",
    "score": 5
  },
  Object {
    "heroName": "Tracer",
    "score": 5
  },
  Object {
    "heroName": "Roadhog",
    "score": 4
  },
  Object {
    "heroName": "McCree",
    "score": 4
  },
  Object {
    "heroName": "Winston",
    "score": 3
  },
  Object {
    "heroName": "Junkrat",
    "score": 3
  },
  Object {
    "heroName": "Reaper",
    "score": 3
  },
  Object {
    "heroName": "Hanzo",
    "score": 2
  },
  Object {
    "heroName": "Genji",
    "score": 2
  },
  Object {
    "heroName": "Soldier: 76",
    "score": 2
  },
  Object {
    "heroName": "Zenyatta",
    "score": 1
  },
  Object {
    "heroName": "Symmetra",
    "score": 1
  },
  Object {
    "heroName": "Ana",
    "score": 1
  },
  Object {
    "heroName": "Zarya",
    "score": 1
  },
  Object {
    "heroName": "Widowmaker",
    "score": 1
  },
  Object {
    "heroName": "Torbjörn",
    "score": 1
  },
  Object {
    "heroName": "Mercy",
    "score": 0
  },
  Object {
    "heroName": "Lúcio",
    "score": 0
  },
  Object {
    "heroName": "Reinhardt",
    "score": 0
  },
  Object {
    "heroName": "D.Va",
    "score": 0
  },
  Object {
    "heroName": "Bastion",
    "score": 0
  },
  Object {
    "heroName": "Pharah",
    "score": 0
  }
]
`;
